src/router.ts

```diff
@@ -0,0 +1,125 @@
+import Router from '@koa/router';
+import { z } from 'zod';
+import { fromZodError } from 'zod-validation-error';
+import zodToJsonSchema from 'zod-to-json-schema';
+import compose = require('koa-compose');
+import { IntrospectionConfig } from './koa';
+import { EndpointImplementation, implementRoute } from './koa-utils';
+import { IntrospectionResponse, OneSchemaDefinition } from './types';
+
+export type RouterEndpointDefinition<Name> = {
+  name: Name;
+  description?: string;
+  request: z.ZodType<any, any, any>;
+  response: z.ZodType<any, any, any>;
+};
+
+type Method = 'GET' | 'DELETE' | 'PUT' | 'POST' | 'PATCH';
+type RoughRoute = `${Method} ${string}`;
+
+type ZodSchema = {
+  [route: string]: RouterEndpointDefinition<string>;
+};
+
+export type OneSchemaRouterConfig<R extends Router<any, any>> = {
+  using: R;
+  introspection: IntrospectionConfig | undefined;
+};
+
+export class OneSchemaRouter<
+  Schema extends ZodSchema,
+  R extends Router<any, any>,
+> {
+  private router: R;
+
+  private constructor(
+    private schema: Schema,
+    { introspection, using: router }: OneSchemaRouterConfig<R>,
+  ) {
+    this.router = router;
+    if (introspection) {
+      router.get(introspection.route, (ctx, next) => {
+        const response: IntrospectionResponse = {
+          serviceVersion: introspection.serviceVersion,
+          schema: convertRouterSchemaToJSONSchemaStyle(this.schema),
+        };
+
+        ctx.body = response;
+        ctx.status = 200;
+        return next();
+      });
+    }
+  }
+
+  static create<R extends Router<any, any>>(
+    config: OneSchemaRouterConfig<R>,
+  ): OneSchemaRouter<{}, R> {
+    return new OneSchemaRouter({}, config);
+  }
+
+  declare<
+    Route extends RoughRoute,
+    Name extends string,
+    Endpoint extends RouterEndpointDefinition<Name>,
+  >(
+    endpoint: Endpoint & { route: Route },
+  ): OneSchemaRouter<Schema & { [route in Route]: Endpoint }, R> {
+    // @ts-expect-error
+    this.schema[endpoint.route] = endpoint;
```

Comment:
Not a huge deal, but looks like this implementation allows for having multiple routes with the same name (looks like the unit tests are doing it e.g. [here](https://github.com/lifeomic/one-schema/pull/52/files#diff-fcbd97d5f830a4e2021a5274c9b972d3901f5bcd5467c9cbef7136cb5d152a58R308)). I think this would only be an issue for generating clients, since there would be ambiguity about which route to associate with a given name.

Not a huge issue though because making a change here would just prevent against potential developer coding errors.

Reply:
That's a _very_ nice catch. I think we should pre-emptively catch that case, and `throw` when someone declares a duplicated name.